### PR TITLE
Dynamically load modules

### PR DESCRIPTION
**Pull request by changlan**
- Compile modules into standalone dynamic libraries
- Load those modules in the `modules` directory at runtime.

### CODE REVIEW

File: core/bessd.cc [hunk 1]
```diff
@@ -360,5 +362,47 @@ bool SetResourceLimit() {
...
+      }
+    }
+  }
+  closedir(dir);
+  return true;
+}
+
+std::string GetCurrentDirectory() {
+  char dest[PATH_MAX];
+  if (readlink("/proc/self/exe", dest, PATH_MAX) == -1) {
```
Comment (barath) [inline core/bessd.cc]: readlink() doesn't guarantee null termination, so dest could either fill up to PATH_MAX or not but still not have a '\0' at the end, and strrchr() won't be able to find the end.

File: core/bessd.cc [hunk 2]
```diff
@@ -360,5 +362,47 @@ bool SetResourceLimit() {
   }
 }
 
+static inline bool has_suffix(const std::string &s, const std::string &suffix) {
```
Comment (barath) [inline core/bessd.cc]: Function comment.

File: core/main.cc [hunk 3]
```diff
@@ -21,6 +21,8 @@ int main(int argc, char *argv[]) {
   google::ParseCommandLineFlags(&argc, &argv, true);
   bess::bessd::ProcessCommandLineArgs();
 
+  bess::bessd::LoadModules(bess::bessd::GetCurrentDirectory() + "modules");
```
Comment (barath) [inline core/main.cc]: If this returns false, should it be fatal?

File: core/bessd.cc [hunk 4]
```diff
@@ -360,5 +362,47 @@ bool SetResourceLimit() {
...
+}
+
+bool LoadModules(const std::string &directory) {
+  DIR *dir = opendir(directory.c_str());
+  if (!dir) {
+    return false;
+  }
+  dirent *entry;
+  while ((entry = readdir(dir)) != nullptr) {
+    if (has_suffix(entry->d_name, ".so")) {
```
Comment (barath) [inline core/bessd.cc]: May also want to check d_type.

**Comment by changlan** [comment]
kat plz? :]

**Comment by sangjinhan** [comment]
^ↀᴥↀ^ meow

**Pull request merged**